### PR TITLE
bash script for creating aws vpc and cleaning the resources by cli

### PR DESCRIPTION
**Pull request by sos24**

### CODE REVIEW

**Review by shtajiryan — CHANGES_REQUESTED** [review]
./createvpc.sh: line 20: trap: ERROR: invalid signal specification
cleaning resources with tag usage and value permanent ...
end of cleaning...

This is the result when "createvpc.sh" is run.
If there are specific steps to be taken for the script to work, let me know, otherwise fixes are needed.

**Comment by sos24** [comment]
> ./createvpc.sh: line 20: trap: ERROR: invalid signal specification cleaning resources with tag usage and value permanent ... end of cleaning...
> 
> This is the result when "createvpc.sh" is run. If there are specific steps to be taken for the script to work, let me know, otherwise fixes are needed.

I have made some changes, it must work now

File: homework-20-12-23/sos-ghazaryan/cleanup.sh [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  # Get all instances
+  instances=$(aws ec2 describe-instances --query 'Reservations[*].Instances[*]' --output json)
+
+  # Filter instances where tag value is not equal to EXCLUDE_TAG_VALUE
+  instances_to_delete=$(echo $instances | jq -r '.[][] | select(.Tags | any(.Key == "'${1}'" and .Value != "'${2}'")).InstanceId')
+
+  # Loop through each instance and terminate it
+  for instance_id in $instances_to_delete; do
+      echo "Terminating instance $instance_id"
+      aws ec2 terminate-instances --instance-ids $instance_id
```
Comment (shtajiryan) [inline homework-20-12-23/sos-ghazaryan/cleanup.sh:30]: The instance takes time to reach terminated state. No resources can be deleted or detached before the instance is in terminated state. There should be a wait/check if the instance has reached terminated state before deleting the rest of the resources.
Currently the cleanup script works if you run it multiple times, first it shuts down the instance and throws errors about resources being in use, if you wait for a minute or two and rerun it, the rest of the resources are deleted.

**Review by shtajiryan — CHANGES_REQUESTED** [review]
Also the resources are not being tagged.

File: homework-20-12-23/sos-ghazaryan/cleanup.sh [hunk 2]
```diff
@@ -27,7 +27,7 @@ function cleanInstancesByTag
 #   Loop through each instance and terminate it
   for instance_id in $instances_to_delete; do
       echo "Terminating instance $instance_id"
-      aws ec2 terminate-instances --instance-ids $instance_id
+      aws ec2 wait terminate-instances --instance-ids $instance_id
```
Comment (shtajiryan) [inline homework-20-12-23/sos-ghazaryan/cleanup.sh]: `aws ec2 wait terminate-instances` does not perform instance deletion, you need to terminate the instance first, and then use this to wait for termination completion

**Review by shtajiryan — CHANGES_REQUESTED** [review]
Creation works correctly.
Cleanup fails on instance deletion.

File: homework-20-12-23/sos-ghazaryan/cleanup.sh (hunk 2 above)
Comment (sos24) [reply homework-20-12-23/sos-ghazaryan/cleanup.sh]: aws ec2 terminate-instances --instance-ids $instance_id with wait was fails reason i change the code like this
`aws ec2 terminate-instances --instance-ids $instance_id
      aws ec2 wait instance-terminated --instance-ids $instance_id`

**Review by shtajiryan — APPROVED** [review]
Works correctly

**Pull request merged**